### PR TITLE
Updates Sample Body for loop checkpoint 2

### PR DESCRIPTION
**Pull request by bdougie**
# What is this?

Every student I have had go through the JavaScript Primer has gotten stuck on Loops and Arrays for considerably longer than any other part of Phase 1.

I have updated the sample code for this loop checkpoint to show an example that is closer to the solution. 
# Issue

Students who have no JavaScript or even generic programming language (more than just HTML) knowledge struggle with these checkpoints. 

We tend to show some bland examples and then push them to solve unrelated problems with little help. I tend to have a lot of extra ad-hoc meetings when students attempt the JavaScript Primer.
# Please review

I am **not** expecting my example to be merged right away, but what I am looking for is a conversation to get a better example for the student to refer to. 
# Question

This is the question the student must solve after seeing this

``` js
Write a function called sumNumbers. This function should:

return the sum of all numbers inside of the array that is passed to the function
return 0 if the array is empty
```

Do we bring up the use of `+=` in any checkpoints prior?

### CODE REVIEW

File: courses/javascript-primer/07-loops/02-for-loops-and-arrays/body.md [hunk 1]
```diff
@@ -25,9 +26,14 @@ Using the `length` property is a common pattern that ensures the loop stops once
...
-"Hi Troy!"
-"Hi Abed!"
-"Hi Annie!"
+"a..."
+"ab..."
+"abc..."
+"abcd..."
+"abcde..."
+"abcdef..."
+"abcdefg..."
```
Comment (mmaxwell) [inline courses/javascript-primer/07-loops/02-for-loops-and-arrays/body.md]: This output is incorrect.

File: courses/javascript-primer/07-loops/02-for-loops-and-arrays/body.md [hunk 2]
```diff
@@ -25,9 +26,14 @@ Using the `length` property is a common pattern that ensures the loop stops once
...
+
 undefined
```
Comment (mmaxwell) [inline courses/javascript-primer/07-loops/02-for-loops-and-arrays/body.md]: Do we talk about why there is a random `undefined` here?

**Comment by cblant20** [comment]
I actually think we shouldn't be showing them the same code in the example and instead just explain the fundamentals. I.e. this is a for loop, this is how you use it, this is when you use it, and the same thing with an array. They should have to use that information to extrapolate the correct solution.

File: courses/javascript-primer/07-loops/02-for-loops-and-arrays/body.md (hunk 2 above)
Comment (bdougie) [inline courses/javascript-primer/07-loops/02-for-loops-and-arrays/body.md]: We don't

**Comment by bdougie** [comment]
@cblant20, this is great feedback. We are definitely not explaining enough of the fundamentals and students are completely lost when they get to this point. This checkpoint is ideally completed in their second week and there is so much they haven't learned or grasped by the time they get to this checkpoint.

I haven't completed every checkpoint in the Primer myself yet, but when asked my student have they seen this `+=`, they couldn't recall what it did or if they saw it. So if we do explain it, we do it so quickly that they forget.  

Any thoughts or examples for what extra information we can add? And where it could be added.

This is my fifth WDT student to get to this point -- I am just trying to figure out if we can do this better.

**Comment by barnes7td** [comment]
@cblant20 This gets down to philosophy of how you present material. I lean the other way. I would rather show them several specific use cases and allow them to build their problem solving mental map through systematic, gradually progress real examples of code. Mainly, because if they can't "extrapolate" yet, which in my experience many can't early on. We have to help them build the mental tools in understanding to do so. 

@bdougie I will have to check if we have introduced the `+=` operator. However, I often tell my students that if that operator is difficult for them to logic with, then write it all out at first. For example.

```
count = count + number;
```

I also would say, that it is OK, that they struggle. This is feedback to the mentor, where the gaps in logic of the student are. Therefore, my goal will never be to remove all struggle. I am not going to rethink fully on this one issue our philosophy of teaching, but I do want to deal with this one checkpoint and hopefully abstract out overall principles of teaching, b/c we are not doing a good job of it in many places in the curriculum.

So my question is, are you changing the code is the lesson, because the exercise needs the `+=` operator and you are saying that it is too disconnected? and also unrealistic for a student on this exercise to know that operator/concept?

**Comment by mmaxwell** [comment]
FWIW, a majority of my students don't solve this problem with `+=` initially. When I ask them to walk me through the logic, and they say something to the effect of "We need to take the number in the array at that spot and add it to the sum", they wind up with something like:

```js
var value = array[i];
sum = sum + value;
```

The variable `value` is unnecessary, but I lean more towards verbosity in the beginning in favor of understanding what's really going on, and then simplifying after the fact. Logically, this code is correct, though it's not the most elegant solution. But I'm looking for progress, not perfection.

I tend to agree with @cblant20 in that it's healthy to teach them to critically think and begin to pull from what they know to fill in the blanks of what they don't, but I think we do have to modify that approach slightly because the truth is, we're limited in the amount of time we get with them. Lest they be blocked and unable to progress, we have to show them more than I would like to.

I will say that as someone with no ruby experience, going through the ruby exercises has been really grueling because it gives you so much that what I'm actually writing is minimal. I'd love to see if there's a way to give the students enough information to be successful, while encouraging them to critically think.

In its current state, I don't think we're set up to do that, but I'd love to explore how we can get there.

File: courses/javascript-primer/07-loops/02-for-loops-and-arrays/body.md (hunk 1 above)
Comment (cblant20) [inline courses/javascript-primer/07-loops/02-for-loops-and-arrays/body.md]: Seconding Matthew here. Can we remove the ellipse?

**Comment by cblant20** [comment]
So I really like this conversation, and I feel like what it's really getting at is simply that we're not sufficiently explaining fundamentals.

Coding is supposed to be hard, so I'll second what the other two guys said about the struggle. Especially at the beginning - they're learning both new languages and new ways to think. So instead of your suggestion of showing them another similar problem, I'd rather teach them more about both loops and arrays, show the concepts in a different context, and explain it in a new way.

**Pull request closed without merging**